libcxx/include/__text_encoding/text_encoding.h

```diff
@@ -0,0 +1,1482 @@
+              *this = __iterator{};
+          } else if (__n < 0) {
+            if ((__data_ + __n) > __text_encoding_data && __data_[__n].__mib_rep == __mib_rep_)
+              __data_ += __n;
+            else
+              *this = __iterator{};
+          }
+        }
+        return *this;
+      }
+
+      _LIBCPP_HIDE_FROM_ABI constexpr __iterator& operator-=(difference_type __n) { return operator+=(-__n); }
+
+      _LIBCPP_HIDE_FROM_ABI constexpr bool operator==(const __iterator& __it) const {
+        return __data_ == __it.__data_ && __it.__mib_rep_ == __mib_rep_;
+      }
+
+      _LIBCPP_HIDE_FROM_ABI constexpr bool operator==(__end_sentinel) const { return !__can_dereference(); }
+
+      _LIBCPP_HIDE_FROM_ABI constexpr auto operator<=>(__iterator __it) const { return __data_ <=> __it.__data_; }
+
+    private:
+      friend struct text_encoding;
+
+      _LIBCPP_HIDE_FROM_ABI constexpr __iterator(const __encoding_data* __enc_d) noexcept
+          : __data_(__enc_d), __mib_rep_(__enc_d ? __enc_d->__mib_rep : 0) {}
+
+      _LIBCPP_HIDE_FROM_ABI bool __can_dereference() const { return __data_ && __data_->__mib_rep == __mib_rep_; }
+
+      // default iterator is a sentinel
+      const __encoding_data* __data_       = nullptr;
+      __encoding_data::__id_rep __mib_rep_ = 0;
+    };
+
+    constexpr __iterator begin() const { return __iterator{__view_data_}; }
+    constexpr __end_sentinel end() const { return __end_sentinel{}; }
+
+  private:
+    const __encoding_data* __view_data_ = nullptr;
+  };
+
+  [[nodiscard]] _LIBCPP_HIDE_FROM_ABI constexpr aliases_view aliases() const _NOEXCEPT {
+    auto __rep = __encoding_rep_ - 1;
+    if (__encoding_rep_->__name[0]) {
+      while (__rep > std::begin(__text_encoding_data) && (__rep--)->__mib_rep == __encoding_rep_->__mib_rep)
+        ;
+    } else {
+      __rep = nullptr;
+    }
+
+    return aliases_view(__rep);
+  }
+
+  _LIBCPP_HIDE_FROM_ABI friend constexpr bool operator==(const text_encoding& __a, const text_encoding& __b) _NOEXCEPT {
+    if (__a.mib() == id::other && __b.mib() == id::other)
+      return __comp_name(__a.__name_, __b.__name_);
+
+    return __a.mib() == __b.mib();
+  }
+
+  _LIBCPP_HIDE_FROM_ABI friend constexpr bool operator==(const text_encoding& __encoding, id __i) _NOEXCEPT {
+    return __encoding.mib() == __i;
+  }
+
+#    if __CHAR_BIT__ == 8
+  [[nodiscard]] _LIBCPP_HIDE_FROM_ABI static consteval text_encoding literal() _NOEXCEPT {
+#      ifdef __GNUC_EXECUTION_CHARSET_NAME
+    return text_encoding(__GNUC_EXECUTION_CHARSET_NAME);
+#      elif defined(__clang_literal_encoding__)
+    return text_encoding(__clang_literal_encoding__);
+#      elif defined(__clang__)
+    return text_encoding(id::UTF8);
+#      else
+    return {};
+#      endif
+  }
+
+  [[nodiscard]] static text_encoding environment();
+
+  template <id __i>
+  [[nodiscard]] _LIBCPP_HIDE_FROM_ABI static bool environment_is() {
+    return environment() == __i;
+  }
```

Comment:
That function intended that it could be optimized - eg for utf8 - such that it would not access / odr-use the data table. But that implementation is fine, especially as a first pass